modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java

```diff
@@ -858,6 +1046,150 @@ private static ParameterMetadata createParameterMetadata(RelDataType parameterRo
+                }
+
+                CompletableFuture<PlanInfo> planFut = rePlanningFut.get();
+                if (planFut != null && !planFut.isDone()) {
+                    // some work still in progress
+                    return;
+                } else {
+                    rePlanningFut.set(null);
+                }
+
+                while (recalculatePlans) {
+                    recalculatePlans = false;
+
+                    for (Entry<CacheKey, CompletableFuture<PlanInfo>> ent : cache.entrySet()) {
+                        if (!ent.getKey().needInvalidate()) {
+                            continue;
+                        }
+
+                        CacheKey key = ent.getKey();
+                        CompletableFuture<PlanInfo> fut = cache.get(key);
+
+                        // can be evicted
+                        if (fut != null) {
+                            assert isCompletedSuccessfully(fut);
+
+                            PlanInfo info = fut.join();
+
+                            assert info.context != null && info.statement != null;
+
+                            int currentCatalogVersion = catalogVersionSupplier.getAsInt();
+
+                            if (currentCatalogVersion == key.catalogVersion()) {
+                                SqlQueryType queryType = info.statement.parsedResult().queryType();
+
+                                if (queryType != SqlQueryType.QUERY && queryType != SqlQueryType.DML) {
+                                    assert false : "Unexpected type: " + queryType;
+                                    continue;
+                                }
+
+                                PlanningContext planningContext = PlanningContext.builder()
+                                        .frameworkConfig(info.context.config())
+                                        .query(info.statement.parsedResult().originalQuery())
+                                        .plannerTimeout(plannerTimeout)
+                                        .catalogVersion(info.context.catalogVersion())
+                                        .defaultSchemaName(info.context.schemaName())
+                                        .parameters(info.context.parameters())
+                                        .explicitTx(info.context.explicitTx())
+                                        .build();
+
+                                CompletableFuture<PlanInfo> newPlanFut =
+                                        prepare.preparePlan(queryType, info.statement.parsedResult, planningContext, key);
+
+                                rePlanningFut.updateAndGet(prev -> prev == null ? newPlanFut : prev.thenCompose(none -> newPlanFut));
+
+                            } else {
+                                key.invalidated();
+                            }
+                        }
+                    }
+                }
+
+            }, PLAN_UPDATER_INITIAL_DELAY, 1_000, TimeUnit.MILLISECONDS);
+        }
+    }
```

Comment:
Key must be immutable. Let's move mutable state to a value.

Reply:
Assume, there is a cache with expiry policy. And two object representing the same key: k1 and k2.
1. I do `cache.put(k1, )`;
2. Then change state `k1.invalidate()`
3. Then k1 is expired, but eviction is lazy... so, entry doesn't actually removed from cache.
4. Then I do `cache.put(k2, )`;
Is there any guarantee that cache entry has k2 or k1?

I bet it depends on implementation. 
The cache may put a new entry with k2, or can just replace entry value keeping k1 in-touched,
just because k1 and k2 are `equal`.

Reply:
done